src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go

```diff
@@ -333,61 +244,247 @@ func (a *StorageMinerActorCode_I) _onSuccessfulPoSt(rt Runtime, postSubmission p
 	TODO() // TODO: validate caller
 
+	// Verify correct PoSt Submission
+	isPoStVerified := a._verifySurprisePoSt(rt, onChainInfo)
+	if !isPoStVerified {
+		// no state transition, just error out and miner should submitSurprisePoSt again
+		// TODO: determine proper error here and error-handling machinery
+		rt.Abort("TODO")
+	}
+
+	return a._onSuccessfulPoSt(rt, onChainInfo)
+
+}
+
+func (a *StorageMinerActorCode_I) _verifySurprisePoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) bool {
+
+	// 1. Check that the miner in question is currently being challenged
 	if !a._isChallenged(rt) {
 		// TODO: determine proper error here and error-handling machinery
 		rt.Abort("cannot SubmitSurprisePoSt when not challenged")
 	}
 
-	// Verify correct PoSt Submission
-	isPoStVerified := a._verifyPoStSubmission(rt, postSubmission)
-	if !isPoStVerified {
-		// no state transition, just error out and miner should submitSurprisePoSt again
+	// to pull in from constants
+	PROVING_PERIOD := block.ChainEpoch(0)
+
+	// 2. Check that the challenge has not expired
+	// Check that miner can still submit (i.e. that the challenge window has not passed)
+	// This will prevent miner from submitting a Surprise PoSt past the challenge period
+	_, st := a.State(rt)
+	if rt.CurrEpoch() < st.ChallengeStatus().LastChallengeEpoch()+PROVING_PERIOD {
+		rt.Abort("cannot SubmitSurprisePoSt late")
+	}
+
+	// A proof must be a valid snark proof with the correct public inputs
+
+	// 3. Get appropriate randomness
+	surpriseRand := rt.Randomness(onChainInfo.PoStEpoch(), 0) //(st.ChallengeStatus().LastChallengeEpoch(), 0)
+
+	// 4. Get public inputs
+	info := st.Info()
+	sectorSize := info.SectorSize()
+
+	postCfg := sector.PoStCfg_I{
+		Type_:        sector.PoStType_SurprisePoSt,
+		SectorSize_:  sectorSize,
+		WindowCount_: info.WindowCount(),
+		Partitions_:  info.SurprisePoStPartitions(),
+	}
+
+	pvInfo := sector.PoStVerifyInfo_I{
+		OnChain_:    onChainInfo,
+		PoStCfg_:    &postCfg,
+		Randomness_: surpriseRand,
+	}
+
+	sdr := filproofs.WinSDRParams(&filproofs.SDRCfg_I{SurprisePoStCfg_: &postCfg})
+
+	// 5. Verify the PoSt Proof
+	isPoStVerified := sdr.VerifySurprisePoSt(&pvInfo)
+	return isPoStVerified
+}
+
+func (a *StorageMinerActorCode_I) VerifyElectionPoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) bool {
+	// TODO: validate caller
+	// the caller MUST be the miner who won the block (who won the block should be callable as a a VM runtime call)
+	// we also need to enforce that this call happens only once per block, OR make it not callable by special privileged messages
+	TODO()
+
+	// 1. Check that the miner in question is currently allowed to run election
+	if !a._canBeElected(rt) {
 		// TODO: determine proper error here and error-handling machinery
-		rt.Abort("TODO")
+		rt.Abort("cannot submit an election proof if challenged or having challenge response failure")
```

Comment:
Return `false`? Or return some status enumeration?

Reply:
Yeah I think we should do a pass for error handling in general. Don't want to do so now.